### PR TITLE
WIP - Fix #363 Change eslint rules and fix any errors

### LINKED ISSUE #363: Fix lint issues on master
**- Do you want to request a *feature* or report a *bug*?**

Bug. See #356 for original discussion.

**- What is the current behavior?**
**- If the current behavior is a bug, please provide the steps to reproduce.**

Execute `npm run lint` on master. Lint fails.

**- What is the expected behavior?**

Existing code passes lint rules. Lint passes.

**- Please mention your node.js, and operating system version.**

Node v7.8.0. macOS Sierra.

### PR DESCRIPTION
**Pull request by josephearl**
**- Summary**

Changes eslint config and fixes the most important issues so that there are no errors when running `npm run lint:js`.

There are still 50 warnings remaining for things I think should probably be addressed in the long run.

**- Description for the changelog**

Added `.eslintignore` to ignore `bin` and `dist` directories.
Ran `npm run lint:js:fix`.
Changed `.eslintrc` to change some errors to warnings and disable some of the more stylistic rules.
Fixed any remaining errors.

**- A picture of a cute animal (not mandatory but encouraged)**

A baby tiger

### CODE REVIEW

File: src/components/stories/Card.js [hunk 1]
```diff
@@ -29,7 +29,7 @@ storiesOf('Card', module)
   )).add('Full width content', () => (
     <Card>
       <img src="https://i.ytimg.com/vi/tntOCGkgt98/maxresdefault.jpg" />
-      <h1>Card & cat</h1>
+      <h1>Card &amp; cat</h1>
```
Comment (josephearl) [inline src/components/stories/Card.js]: I think this change is unnecessary, will revert

**Comment by erquhart** [comment]
@josephearl thank you for doing this, long time coming! Just a few small changes for what you have so far, I'll pull it down and take a look at the remaining items you mentioned.

File: src/backends/github/API.js [hunk 2]
```diff
@@ -486,15 +486,15 @@ export default class API {
             }
           }
         }
-        for (filename in fileTree) {
+        Object.keys(fileTree).forEach((filename) => {
           fileOrDir = fileTree[filename];
-          if (added[filename]) { continue; }
+          if (added[filename]) { return; }
```
Comment (josephearl) [inline src/backends/github/API.js]: Done

File: src/components/Widgets/ControlHOC.js [hunk 3]
```diff
@@ -44,7 +44,7 @@ class ControlHOC extends Component {
     if (isRequired && (
       value === null ||
       value === undefined ||
-      (value.hasOwnProperty('length') && value.length === 0) ||
+      (Object.prototype.hasOwnProperty.call(value, 'length') && value.length === 0) ||
```
Comment (josephearl) [inline src/components/Widgets/ControlHOC.js]: Done

File: src/lib/netlify-auth.js [hunk 4]
```diff
@@ -36,7 +36,7 @@ class Authenticator {
 
   handshakeCallback(options, cb) {
     const fn = (e) => {
-      if (e.data === ('authorizing:' + options.provider) && e.origin === this.base_url) {
+      if (e.data === (`authorizing:${ options.provider }`) && e.origin === this.base_url) {
```
Comment (josephearl) [inline src/lib/netlify-auth.js]: Done

File: src/components/Widgets/ControlHOC.js [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import React, { Component, PropTypes } from 'react';
 import ImmutablePropTypes from "react-immutable-proptypes";
+import _ from "lodash";
```
Comment (josephearl) [inline src/components/Widgets/ControlHOC.js]: or should I `import { has } from "lodash";`?

Comment (erquhart) [inline src/components/Widgets/ControlHOC.js]: Yeah just take the method you're using.

Comment (josephearl) [inline src/components/Widgets/ControlHOC.js]: Done

**Comment by erquhart** [comment]
LGTM - @Benaiah please have a look and we can merge this ahead of your upcoming formatting PR and run Prettier on top of it.

File: .eslintrc [hunk 6]
```diff
@@ -2,6 +2,36 @@
...
+    "react/jsx-no-bind": "warn",
+    "import/prefer-default-export": "off",
+    "import/imports-first": [ "warn", "DISABLE-absolute-first" ],
+    "max-len": "off",
+    "no-alert": "off",
+    "no-console": "off",
+    "no-cond-assign": "off",
+    "no-continue": "off",
+    "no-else-return": "off",
+    "no-shadow": "off",
```
Comment (erquhart) [inline .eslintrc]: Yeah I actually like this rule, as annoying as it is with react-redux `connect`. I'd be okay with turning it off if we had a low priority ticket for turning it back on.

**Comment by erquhart** [comment]
Awesome review @Benaiah.

My plan was to take the shortest path to passing linting, and then bring rules back as we move forward, but I think the path you laid out is doable, especially if we leave `no-shadow` out for now. That said, if we're going to up the workload, let's make this PR a group effort - pushing that last commit from your branch, or at least the rule changes and related code updates, would be a good move. But that's just based on a quick glance.

Two thoughts:
* I totally agree with sticking closely to eslint-config-netlify, worth the effort
* Only leave warnings up for things that we're willing to fix, otherwise it's just noise

File: .eslintrc [hunk 7]
```diff
@@ -2,6 +2,36 @@
...
+    "no-unneeded-ternary": "warn",
+    "react/no-danger": "warn",
+    "react/no-find-dom-node": "warn",
+    "react/no-unescaped-entities": "warn",
+    "react/jsx-no-bind": "warn",
+    "import/prefer-default-export": "off",
+    "import/imports-first": [ "warn", "DISABLE-absolute-first" ],
+    "max-len": "off",
+    "no-alert": "off",
+    "no-console": "off",
```
Comment (calavera) [inline .eslintrc]: I'd rather turn this on.

File: .eslintrc [hunk 8]
```diff
@@ -2,6 +2,36 @@
...
+    "no-unused-expressions": "warn",
+    "no-unneeded-ternary": "warn",
+    "react/no-danger": "warn",
+    "react/no-find-dom-node": "warn",
+    "react/no-unescaped-entities": "warn",
+    "react/jsx-no-bind": "warn",
+    "import/prefer-default-export": "off",
+    "import/imports-first": [ "warn", "DISABLE-absolute-first" ],
+    "max-len": "off",
+    "no-alert": "off",
```
Comment (calavera) [inline .eslintrc]: agreed

File: .eslintrc [hunk 9]
```diff
@@ -2,6 +2,36 @@
...
+    "no-console": "off",
+    "no-cond-assign": "off",
+    "no-continue": "off",
+    "no-else-return": "off",
```
Comment (calavera) [inline .eslintrc]: agreed

File: .eslintrc [hunk 10]
```diff
@@ -2,6 +2,36 @@
...
+    "no-shadow": "off",
+    "no-unused-vars": "off",
```
Comment (calavera) [inline .eslintrc]: I agree, I would rather not turn this rule off.

**Comment by Benaiah** [comment]
@erquhart @josephearl would you like me to make a new PR with the history stemming from this one, or just push my commits to this PR?

**Comment by josephearl** [comment]
@Benaiah I'd say push to this one

**Comment by erquhart** [comment]
@Benaiah what's the status on this?

**Comment by erquhart** [comment]
@Benaiah I'm tempted to close this and let someone re-tackle from the current codebase. Please close if you agree.

**Comment by erquhart** [comment]
Closing due to the number of requested changes pending, and this PR being so far out of date.

**Pull request closed without merging**